src/plan.rs

```diff
@@ -0,0 +1,139 @@
+use std::{
+    cmp::Ordering,
+    collections::{BinaryHeap, HashMap},
+};
+
+pub struct Id {
+    id: usize,
+}
+
+pub struct Plan<T> {
+    pub time: f64,
+    pub data: T,
+}
+
+#[derive(PartialEq, Debug)]
+pub struct Record {
+    pub time: f64,
+    id: usize,
+}
+
+impl Eq for Record {}
+
+impl PartialOrd for Record {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Record {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let time_ordering = self.time.partial_cmp(&other.time).unwrap().reverse();
+        if time_ordering == Ordering::Equal {
+            // Break time ties in order of plan id
+            self.id.cmp(&other.id).reverse()
+        } else {
+            time_ordering
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Queue<T> {
+    queue: BinaryHeap<Record>,
+    data_map: HashMap<usize, T>,
+    plan_counter: usize,
+}
+
+impl<T> Default for Queue<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T> Queue<T> {
+    #[must_use]
+    pub fn new() -> Queue<T> {
+        Queue {
+            queue: BinaryHeap::new(),
+            data_map: HashMap::new(),
+            plan_counter: 0,
+        }
+    }
+
+    pub fn add_plan(&mut self, time: f64, data: T) -> Id {
+        // Add plan to queue, store data, and increment counter
+        let id = self.plan_counter;
+        self.queue.push(Record { time, id });
+        self.data_map.insert(id, data);
+        self.plan_counter += 1;
+        Id { id }
+    }
+
+    /// # Panics
+    ///
+    /// This function panics if you cancel a plan which has already
+    /// been cancelled or executed.
+    pub fn cancel_plan(&mut self, id: &Id) {
+        self.data_map.remove(&id.id).expect("Plan does not exist");
```

Comment:
I would add a comment, as in:

`Delete the plan from the map, but leave it in the queue. It will be skipped
when we get to the deadline`.